core/engine_test.go

```diff
@@ -241,76 +243,82 @@ func TestEngineOutput(t *testing.T) {
 	}
 }
 
-func TestEngine_processSamples(t *testing.T) {
+func TestEngineProcessSamplesOfMetric(t *testing.T) {
 	t.Parallel()
+
+	// Arrange
 	metric := stats.New("my_metric", stats.Gauge)
+	e, _, wait := newTestEngine(t, nil, nil, nil, lib.Options{})
+	defer wait()
 
-	t.Run("metric", func(t *testing.T) {
-		t.Parallel()
-		e, _, wait := newTestEngine(t, nil, nil, nil, lib.Options{})
-		defer wait()
+	// Act
```

Comment:
I'm against having the code with a lot of these comments, if it isn't clear where the arrange-act-assert state operations start and end then it probably means that the test needs to be refactored again.

Reply:
Me too, these comments don't really add anything besides noise for me :confused:

Reply:
I wholeheartedly disagree with you folks, but I don't have a strong enough opinion about it to argue 😄 

To expand on my view on this: tests are also documentation. They need to be easy to read/parse, structured, and make their intent and the behavior they outline explicit. I personally had a hard time reading through them and understanding them; as it required me quite a bit of overhead. One of the main reasons, I found, was that it was actually not clear to me what is set up code (arrange), what is actually being tested (act), and what is the expected output state (assert).

Reply:
Googling "AAA testing" leads me to
> Abdominal aortic aneurysm (AAA) screening is a way of checking if there's a bulge or swelling in the aorta

:sweat_smile: The second result is https://medium.com/@pjbgf/title-testing-code-ocd-and-the-aaa-pattern-df453975ab80 though - do you have a better introduction to the concept? Because I think that's written by someone who likes the concept, like you, but also has the same opinion as @codebien and me about the explicit comments:
> Test code, in the same way as production code, is supposed to be clean and self-explanatory. Preempting what you are going to do just adds unnecessary clutter. You can still have a clear distinction of the three steps by using spaces and indentation. Which are ultimately the key features the compiler provides you to structure your code.
> It does look silly having a comment preceding each paragraph in this text, it tires the reader, as it has to constantly go through the obvious (or skip it). So why would you do that with your code?

Couldn't have written it better myself :sweat_smile: :+1: for refactoring tests to be cleaner and easier to read, ideally self-explanatory. The same as any other piece of code. :-1: for comments that should be obvious from the context.

Reply:
Sorry I used an acronym, I'm generally the one chasing people telling them not to do it, and here I go losing at my own game 🟥 ...

AAA stands for Arrange, Act, Assert. It's a very common pattern in software testing, specifically because it helps outline and make the natural flow of any unit tests explicit: arrange some input, act upon them, assert the produced output.

I don't mean to play the "someone is wrong on the internet" card, but I found the article you point at quite judgmental, and left me disagreeing rather than convinced. Here a few counter examples I would like to point you too, from [Microsoft](https://docs.microsoft.com/en-us/visualstudio/test/unit-test-basics?view=vs-2022), a Manning book about [Unit testing](https://freecontent.manning.com/making-better-unit-tests-part-1-the-aaa-pattern/), or a [blog post](https://automationpanda.com/2020/07/07/arrange-act-assert-a-pattern-for-writing-good-tests/) I often end up stumbling upon when I need a refresher.

Here, I end up arguing about it, although I wanted to actually avoid that. All I can say is that I've been introduced to this technique years ago by a colleague. I was really skeptical at first, but I never went back once I started. I spend 90% reading code rather than writing it, and anything that makes my life "easier" as a reader is a gain from my perspective. It took me a lot of effort to make sense out of these tests, I found them (too) long, complicated, and it was unclear to me at first glance what was actually setting up code, actually "act" code, and assertion code. You might have a different experience, but that was mine 😇 

Now I acknowledge everyone has a different experience, a different brain, and that's exactly why I'm not especially trying to convince you folks. What becomes apparent, though, is that I would like to have some document somewhere where we gather, discuss, and decide on best practices we follow on this repository to be able to quickly sort out these disagreements in the future 😄

Reply:
>  It took me a lot of effort to make sense out of these tests, I found them (too) long, complicated, and it was unclear to me at first glance what was actually setting up code, actually "act" code, and assertion code. You might have a different experience, but that was mine innocent

I don't disagree with this part, a lot of the tests can definitely be improved, **substantially** :sweat_smile: I don't even disagree that tests should be split into these 3 sections whenever possible, that's the natural flow for most unit tests. However, I disagree that we should add these `// Arrange`, `// Act`, `// Assert` comments in the tests. Instead, I think the tests should be refactored to be so clean and easy to read that we don't need such comments, i.e. in such a way that the sections are obvious without them :slightly_smiling_face: